### PR TITLE
Change pip downloading sources to pypi instead of via pip itself

### PR DESCRIPTION
**Pull request by davidlange6**
33-xx version of #174

### CODE REVIEW

File: cmsBuild [hunk 1]
```diff
@@ -619,6 +621,31 @@ def downloadPip(source, dest, options):
...
+    if isSourceDownload:
+        log("Looking for sources at https://pypi.org/pypi/"+pack.split('=')[0]+"/json")
+        fj=urlopen("https://pypi.org/pypi/"+pack.split('=')[0]+"/json")
+        data=json.load(fj)
+        url=None
+        if "releases" in data and pack.split('=')[2] in data["releases"]:
+            for file in data["releases"][pack.split('=')[2]]:
+                if file["packagetype"] == "sdist":
+                    url=file["url"]
+        if url is not None:
```
Comment (mrodozov) [inline cmsBuild:636]:
```suggestion
        if url:
```

**Comment by davidlange6** [comment]
hi @mrodozov @smuzaffar -aside from the small fix suggested above, any other comments on this?

**Comment by mrodozov** [comment]
it says draft so I thought you still have work on it

**Comment by davidlange6** [comment]
I can remove draft indeed - It works at the level I’ve tested it anyway:)

**Comment by smuzaffar** [comment]
@davidlange6 , I am also testing it locally by forcing re-build of all pip based packages and re-download of sources . I will get back once testing is done

**Comment by davidlange6** [comment]
Cool - thanks - btw, how does one force a redownload of spruces? I failed to figure that out...

**Comment by smuzaffar** [comment]
you need to change the url checksum. I did it by changing `cmsBuild` itself [a] but you can also do it by adding a dummy parameter to pip source url in cmsdist/build-with-pip.file e.g.
```
%define source0 pip://%{pip_name}/%{realversion}?pip_options=%{PipDownloadOptions}&foo=bar&pip=%{pip}&output=/%{pkgsource}
```

[a] 
```
@@ -790,6 +791,9 @@ def download(source, dest, options, pkg=None):
     noCmssdtCache = True if 'no-cmssdt-cache=1' in source else False
     source = fixUrl(source)
     checksum = getUrlChecksum(source)
+    print("SOURCE: %s" % source)
+    if source.startswith('pip://'):
+      checksum = getUrlChecksum(source+"/foo")

     # Syntactic sugar to allow the following urls for tag collector:
```

**Comment by smuzaffar** [comment]
+externals

changes looks good. These are the packages [a] for which the new code was hit. The issue with `boost-histogram==0.13` is due to wrong version in cmsdist. It should have been 0.13.0. 


[a]
- flit-core==3.1.0 = https://files.pythonhosted.org/packages/4c/8f/bed80c03f71cb3a2935882f391b53d2510c359191e5e0361650fa02d1365/flit_core-3.1.0.tar.gz
- setuptools-scm==3.5.0 = https://files.pythonhosted.org/packages/b2/f7/60a645aae001a2e06cf4b8db2fba9d9f36b8fd378f10647e3e218b61b74b/setuptools_scm-3.5.0.tar.gz
- importlib-metadata==2.1.1 = https://files.pythonhosted.org/packages/40/a2/0d11187bdde4237a2f5413f3dd47bdce4d198215392cf6c69945e89c6797/importlib_metadata-2.1.1.tar.gz
- simplegeneric==0.8.1 = https://files.pythonhosted.org/packages/3d/57/4d9c9e3ae9a255cd4e1106bb57e24056d3d0709fc01b2e3e345898e49d5b/simplegeneric-0.8.1.zip
- pillow==8.1.1 = https://files.pythonhosted.org/packages/a6/24/1346f8c70dae5daf58e22435a1f1f4696682b4f85321eb4d18ca1d81c0c2/Pillow-8.1.1.tar.gz
- distlib==0.3.1 = https://files.pythonhosted.org/packages/2f/83/1eba07997b8ba58d92b3e51445d5bf36f9fba9cb8166bcae99b9c3464841/distlib-0.3.1.zip
- Click==7.1.2 = https://files.pythonhosted.org/packages/27/6f/be940c8b1f1d69daceeb0032fee6c34d7bd70e3e649ccac0951500b4720e/click-7.1.2.tar.gz
- importlib-resources==3.3.1 = https://files.pythonhosted.org/packages/e1/bf/b40e6ecb611fa3be5b0ecae7b37ae1fa0d8df3146477ec0a554e8f62c976/importlib_resources-3.3.1.tar.gz
- keras-preprocessing==1.1.2 = https://files.pythonhosted.org/packages/5e/f1/b44337faca48874333769a29398fe4666686733c8880aa160b9fd5dfe600/Keras_Preprocessing-1.1.2.tar.gz
- keras-applications==1.0.8 = https://files.pythonhosted.org/packages/21/56/4bcec5a8d9503a87e58e814c4e32ac2b32c37c685672c30bc8c54c6e478a/Keras_Applications-1.0.8.tar.gz
- pillow==6.2.2 = https://files.pythonhosted.org/packages/b3/d0/a20d8440b71adfbf133452d4f6e0fe80de2df7c2578c9b498fb812083383/Pillow-6.2.2.tar.gz
- typed-ast==1.4.2 = https://files.pythonhosted.org/packages/36/8c/efd8ffe7d242cd389632a11cbc6ce596de49b46ece22760a67b742534368/typed_ast-1.4.2.tar.gz
- networkx==2.2 = https://files.pythonhosted.org/packages/f3/f4/7e20ef40b118478191cec0b58c3192f822cace858c19505c7670961b76b2/networkx-2.2.zip
- cachecontrol==0.12.6 = https://files.pythonhosted.org/packages/44/b9/9a1d4349824ae14de4052e84802a0c372fff1fd2bd6668f7e9efe91ac11d/CacheControl-0.12.6.tar.gz
- boost-histogram==0.13 = None
- sqlalchemy==1.3.11 = https://files.pythonhosted.org/packages/34/5c/0e1d7ad0ca52544bb12f9cb8d5cc454af45821c92160ffedd38db0a317f6/SQLAlchemy-1.3.11.tar.gz
- hepdata-lib==0.4.1 = https://files.pythonhosted.org/packages/d1/7a/fc94192bd965fe4802ce232a3e0efa812b2195ca0f30afe0ee1bc51abfd5/hepdata_lib-0.4.1.tar.gz
- cx-Oracle==7.3.0 = https://files.pythonhosted.org/packages/5e/e3/cfd4f413f8d47dd2aee09273d9bb3ee6aa7384e0a04e191d703c2199eb93/cx_Oracle-7.3.0.tar.gz

File: cmsBuild (hunk 1 above)
Comment (davidlange6) [reply cmsBuild:636]: it seems is not None is the more standard formulation.. "Comparisons to singletons like None should always be done with is or is not, never the equality operators."

**Pull request merged**